### ISSUE TITLE
TLS routing connection upgrade using standard WebSocket

### ISSUE THREAD

**Issue by greedy52** [gravitational/teleport#30493]
What would you like Teleport to do?
TLS routing connection upgrade using standard WebSocket `websocket`, in addition to `alpn` and `alpn-ping`.

What problem does this solve?
Currently, TLS routing uses custom connection upgrades for L7 LB/reverse proxies. But many reverse proxies only support connection upgrade type as `websocket`.

If a workaround exists, please include it.
No

**Comment by webvictim** [comment]
For a little extra context, commercial proxies like Cloudflare/Akamai/Zscaler etc often only allow HTTPS traffic and websocket upgrades, they deny all other connection upgrades. A lot of our potential customers will be behind these kinds of proxies and if we can use straight HTTPS/websocket connections, we will increase compatibility. Cloudflare in particular is mentioned very often by community users.

An example of the errors trying to log into Teleport with `tsh login` behind a Cloudflare proxy (that does not allow custom connection upgrades):
```
ubuntu@ip-172-31-30-140:~$ TELEPORT_LOGIN_BIND_ADDR=127.0.0.1:54321 tsh -d login --proxy=teleport.example.com
2023-08-15T15:15:50Z DEBU [TSH]       Web proxy port was not set. Attempting to detect port number to use. tsh/tsh.go:3826
2023-08-15T15:15:50Z DEBU [TSH]       Resolving default proxy port (insecure: false) tsh/resolve_default_addr.go:111
2023-08-15T15:15:50Z DEBU [TSH]       Trying teleport.example.com:3080... tsh/resolve_default_addr.go:99
2023-08-15T15:15:50Z DEBU [TSH]       Trying teleport.example.com:443... tsh/resolve_default_addr.go:99
2023-08-15T15:15:50Z DEBU [TSH]       Address teleport.example.com:443 succeeded. Selected as canonical proxy address tsh/resolve_default_addr.go:191
2023-08-15T15:15:50Z DEBU [TSH]       Waiting for all in-flight proxy address tests to finish tsh/resolve_default_addr.go:139
2023-08-15T15:15:50Z DEBU [TSH]       Proxy address test failed error:[Get "https://teleport.example.com:3080/webapi/ping": context canceled] tsh/resolve_default_addr.go:63
2023-08-15T15:15:50Z INFO [CLIENT]    No teleport login given. defaulting to ubuntu client/api.go:1033
2023-08-15T15:15:50Z INFO [CLIENT]    no host login given. defaulting to ubuntu client/api.go:1043
2023-08-15T15:15:50Z ERRO [CLIENT]    [KEY AGENT] Unable to connect to SSH agent on socket: "". client/api.go:4293
2023-08-15T15:15:50Z DEBU [TSH]       Pinging the proxy to fetch listening addresses for non-web ports. tsh/tsh.go:3315
2023-08-15T15:15:50Z DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:4252
2023-08-15T15:15:50Z DEBU             Attempting GET teleport.example.com:443/webapi/ping webclient/webclient.go:129
2023-08-15T15:15:50Z DEBU             ALPN connection upgrade required for "teleport.example.com:443": true. client/alpn_conn_upgrade.go:95
2023-08-15T15:15:50Z DEBU [CLIENT]    Attempting to login with a new RSA private key. client/api.go:3636
2023-08-15T15:15:51Z DEBU [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client/api.go:4252
2023-08-15T15:15:51Z DEBU [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false, extraHeaders=map[]) client/weblogin.go:308
2023-08-15T15:15:51Z DEBU [CLIENT]    Binding to 127.0.0.1:54321. client/redirect.go:144
2023-08-15T15:15:51Z INFO [CLIENT]    Waiting for response at: http://127.0.0.1:54321/. client/redirect.go:157
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:54321/38f4f0dd-dd98-4aae-89bd-7f370135beb2
2023-08-15T15:15:58Z DEBU [CLIENT]    Got response from browser. client/weblogin.go:396
2023-08-15T15:15:58Z DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-15 23:14:15 +0000 UTC". client/client_store.go:106
2023-08-15T15:15:58Z DEBU [KEYAGENT]  Deleting obsolete stored key with index {ProxyHost:teleport.example.com Username:webvictim ClusterName:teleport.example.com}. client/keyagent.go:527
2023-08-15T15:15:58Z DEBU [KEYSTORE]  Adding known host teleport.example.com with proxy teleport.example.com client/trusted_certs_store.go:393
2023-08-15T15:15:58Z INFO [KEYAGENT]  Loading SSH key for user "webvictim" and cluster "teleport.example.com". client/keyagent.go:195
2023-08-15T15:15:58Z DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-15 23:15:58 +0000 UTC". client/client_store.go:106
2023-08-15T15:15:58Z DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-15 23:15:58 +0000 UTC". client/client_store.go:106
2023-08-15T15:15:58Z INFO [CLIENT]    Connecting to proxy=teleport.example.com:443 login="gus" using TLS Routing client/api.go:3040
2023-08-15T15:15:58Z DEBU [KEYSTORE]  Teleport TLS certificate valid until "2023-08-15 23:15:58 +0000 UTC". client/client_store.go:106
2023-08-15T15:15:58Z DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:195
2023-08-15T15:15:58Z DEBU             ALPN connection upgrade for teleport.example.com:443. client/alpn_conn_upgrade.go:187

ERROR REPORT:
Original Error: tls.RecordHeaderError tls: first record does not look like a TLS handshake
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/client/alpn.go:148 github.com/gravitational/teleport/api/client.(*ALPNDialer).DialContext
	github.com/gravitational/teleport/lib/utils/proxy/proxy.go:71 github.com/gravitational/teleport/lib/utils/proxy.directDial.DialTimeout
	github.com/gravitational/teleport/lib/utils/proxy/proxy.go:58 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial
	github.com/gravitational/teleport/lib/client/api.go:3122 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper
	github.com/gravitational/teleport/lib/client/api.go:3041 github.com/gravitational/teleport/lib/client.makeProxySSHClient
	github.com/gravitational/teleport/lib/client/api.go:2988 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy
	github.com/gravitational/teleport/lib/client/api.go:2965 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1
	runtime/asm_amd64.s:1598 runtime.goexit
User Message: Unable to connect to ssh proxy at teleport.example.com:443. Confirm connectivity and availability.
	tls: first record does not look like a TLS handshake
```

**Comment by greedy52** [comment]
A very very raw POC using WebSocket. Tested against AWS ALB, but I don't have a setup to test where it requires strict websocket upgrade.
https://github.com/gravitational/teleport/compare/master...STeve/30493/poc

**Comment by greedy52** [comment]
For testing purpose, see:
- #31202

**Comment by zmb3** [comment]
@greedy52 can this be closed?

**Comment by greedy52** [comment]
closed by #36343, released in [v15.1.0](https://github.com/gravitational/teleport/releases/tag/v15.1.0)
